### PR TITLE
Cleanup up and Fixing component creation and deletion

### PR DESCRIPTION
**Pull request by sithhell**
This patch contains various fixes related to component registration, creation
and deletion.

The main motivation was to allow direct creation of components without going
through a factory function, disallowing perfect forwarding for local component
creation.

The component registration has been cleaned up, the factory class for creation
and deletion of components has been removed, and the component type registration
has been moved to the component_registry.

The component deletion has been moved out of the runtime_support server, since
no factory is required anymore.

Various other cleanups have been performed leading to an overall reduction in
binary size and removal of various different duplicate code while preserving
existing functionality.
    
Fixes include:
 - The functionality of component instance counters has been fixed for
          local_new
 - The deletion of components now correctly takes migrated components
          into account
 - The component heap instances are now ensured to live in single module only
          avoiding freeing memory that has been allocated by a different module.

Other flyby changes include some cleanup of the runtime_support server and the component interfaces.

### CODE REVIEW

File: hpx/runtime/parcelset/detail/call_for_each.hpp [hunk 1]
```diff
@@ -45,12 +45,13 @@ namespace hpx { namespace parcelset
...
 
-            void operator()(boost::system::error_code const& e) const
+            void operator()(boost::system::error_code const& e)
             {
                 HPX_ASSERT(parcels_.size() == handlers_.size());
                 for(std::size_t i = 0; i < parcels_.size(); ++i)
                 {
                     handlers_[i](e, parcels_[i]);
+                    handlers_[i].reset();
                 }
```
Comment (hkaiser) [inline hpx/runtime/parcelset/detail/call_for_each.hpp]: Is this change related to this PR?

File: src/runtime/components/component_type.cpp [hunk 2]
```diff
@@ -64,5 +151,43 @@ namespace hpx { namespace components
...
+            if (base_name)
+            {
+                // NOTE: This assumes that the derived component is loaded.
+                if (base_type == component_invalid)
+                {
+                    base_type = agas_client.get_component_id(base_name);
+                }
+                type = derived_component_type(type, base_type);
+//                 components::enabled(components::get_derived_type(type)) = enabled;
+//                 components::enabled(components::get_base_type(type)) = enabled;
```
Comment (hkaiser) [inline src/runtime/components/component_type.cpp]: Can this be removed?

File: src/runtime/components/server/destroy_component.cpp [hunk 3]
```diff
@@ -4,20 +4,72 @@
...
 {
     ///////////////////////////////////////////////////////////////////////////
     void destroy_component(naming::gid_type const& gid,
-        naming::address const& addr, error_code& ec)
+        naming::address const& addr_)
     {
-        using components::stubs::runtime_support;
-        agas::gva g (addr.locality_, addr.type_, 1, addr.address_);
-        runtime_support::free_component_sync(g, gid, 1);
+        naming::address addr(addr_);
```
Comment (hkaiser) [inline src/runtime/components/server/destroy_component.cpp]: Why do you need a local copy?

File: hpx/runtime/components/component_type.hpp [hunk 4]
```diff
@@ -151,18 +166,61 @@ namespace hpx { namespace components
...
+//             {
+//                 type = detail::get_agas_component_type(
+//                     get_component_name<Component>(),
+//                     get_component_base_name<Component>());
+//                 traits::component_type_database<Component>::set(type);
+//             }
+//
+//             return type;
+//         }
+//     }
```
[the remark this answers is not in the document]
Comment (sithhell) [reply hpx/runtime/components/component_type.hpp]: Yes.

File: hpx/runtime/components/server/create_component.hpp [hunk 5]
```diff
@@ -8,102 +8,104 @@
...
+                get_component_type_name(type) + ")");
+            return naming::invalid_gid;
         }
 
-        Component::destroy(c, count);
-
-        HPX_THROW_EXCEPTION(hpx::unknown_component_address,
-            "create<Component>",
-            "can't assign global id");
+        void *storage = component_heap<Component>().alloc(1);
```
[the remark this answers is not in the document]
Comment (sithhell) [reply hpx/runtime/components/server/create_component.hpp]: It might throw. In the case of an exception, it is propagated up to the caller. This is the same behavior as when the `Component` ctor throws.

File: hpx/runtime/components/server/create_component.hpp [hunk 6]
```diff
@@ -8,102 +8,104 @@
...
+        try
+        {
+            c = new(storage) Component(std::forward<Ts>(ts)...);
         }
         catch(...)
         {
-            Component::heap_type::free(cv, 1); //-V107
+            component_heap<Component>().free(c, 1);
             throw;
+            return naming::invalid_gid;
```
[the remark this answers is not in the document]
Comment (sithhell) [reply hpx/runtime/components/server/create_component.hpp]: Good catch.

File: src/runtime/components/server/destroy_component.cpp (hunk 3 above)
Comment (sithhell) [reply src/runtime/components/server/destroy_component.cpp]: The local copy is needed because `agas::is_local_address_cached` requires an lvalue.

File: hpx/runtime/parcelset/detail/call_for_each.hpp (hunk 1 above)
Comment (sithhell) [reply hpx/runtime/parcelset/detail/call_for_each.hpp]: Not directly... Not sure why I included it. Potentially, it leads to resources being freed in a timelier fashion. I could add it in a different PR if you wish.

**Comment by sithhell** [comment]
The comments have been addressed.
